st2common/st2common/util/green/shell.py

```diff
@@ -70,54 +72,98 @@ def run_command(cmd, stdin=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+
 
     :rtype: ``tuple`` (exit_code, stdout, stderr, timed_out)
     """
-    LOG.debug("Entering run_command.")
+    LOG.debug('Entering run_command.')
     assert isinstance(cmd, (list, tuple) + six.string_types)
 
-    LOG.debug("Setting env if not set.")
+    if (read_stdout_func and not read_stderr_func) or (read_stderr_func and not read_stdout_func):
+        raise ValueError('Both read_stdout_func and read_stderr_func arguments need '
+                         'to be provided.')
+
+    if read_stdout_func and not (read_stdout_buffer or read_stderr_buffer):
+        raise ValueError('read_stdout_buffer and read_stderr_buffer arguments need to be provided '
+                         'when read_stdout_func is provided')
+
+    LOG.debug('Setting env if not set.')
     if not env:
         env = os.environ.copy()
 
     # Note: We are using eventlet friendly implementation of subprocess
     # which uses GreenPipe so it doesn't block
-    LOG.debug("Creating subprocess.")
+    LOG.debug('Creating subprocess.')
     process = subprocess.Popen(args=cmd, stdin=stdin, stdout=stdout, stderr=stderr,
                                env=env, cwd=cwd, shell=shell, preexec_fn=preexec_func)
 
+    if read_stdout_func:
+        LOG.debug('Spawning read_stdout_func function')
+        read_stdout_thread = eventlet.spawn(read_stdout_func, process.stdout, read_stdout_buffer)
+
+    if read_stderr_func:
+        LOG.debug('Spawning read_stderr_func function')
+        read_stderr_thread = eventlet.spawn(read_stderr_func, process.stderr, read_stderr_buffer)
+
     def on_timeout_expired(timeout):
         global timed_out
 
         try:
-            LOG.debug("Starting process wait.")
+            LOG.debug('Starting process wait.')
             process.wait(timeout=timeout)
         except subprocess.TimeoutExpired:
             # Command has timed out, kill the process and propagate the error.
             # Note: We explicitly set the returncode to indicate the timeout.
-            LOG.debug("Timeout.")
+            LOG.debug('Timeout.')
             process.returncode = TIMEOUT_EXIT_CODE
 
             if kill_func:
-                LOG.debug("Kill_func.")
+                LOG.debug('Kill_func.')
                 kill_func(process=process)
             else:
-                LOG.debug("Kill process.")
+                LOG.debug('Kill process.')
                 process.kill()
 
-    LOG.debug("Setting up process and callback.")
+            if read_stdout_func and read_stderr_func:
+                LOG.debug('Killing read_stdout_thread and read_stderr_thread')
+                read_stdout_thread.kill()
+                read_stderr_thread.kill()
+
+    LOG.debug('Setting up process and callback.')
     timeout_thread = eventlet.spawn(on_timeout_expired, timeout)
-    LOG.debug("Attaching to process.")
-    stdout, stderr = process.communicate()
+    LOG.debug('Attaching to process.')
+
+    if read_stderr_func and read_stderr_func:
```

Comment:
There's two read_stderr_func here.  Should it be `if read_stdout_func and read_stderr_func`?